src/yang/ietf-bgp-capabilities.yang

```diff
@@ -0,0 +1,339 @@
+      "RFC XXXX, BGP Model for Service Provider Network.";
+  }
+
+  grouping bgp-capabilities-common {
+    description
+      "BGP Capabilities that are used commonly in by the
+       advertised-capabilities and received-capabilities lists in the
+       bgp-capabilities grouping.";
+
+    leaf code {
+      type uint8;
+      description
+        "BGP Capability Code";
+      reference
+        "RFC 5492: Capabilities Advertisement with BGP-4, Section
+         4."; }
+    leaf index {
+      type uint8;
+      description
+        "Multiple BGP Capabilities with a given Capability Code may
+         be advertised in a BGP OPEN message.  This index permits
+         multiple such Capabilities to be represented in the YANG
+         model.  It is RECOMMENDED that this index start at one (1)
+         and increase by one for each additional instance of the
+         Capability.";
+      reference
+        "RFC 5492: Capabilities Advertisement with BGP-4, Section
+         4."; }
+    leaf name {
+      type identityref {
+        base bt:bgp-capability;
+      }
+      description
+        "When known, name carries the bgp-capability identity for the
+         AFI/SAFI combination as used in the BGP YANG modules.";
+    }
+
+    container value {
+      description
+        "Some BGP Capabilities carry a Capability-specific Capability
+         Value.";
+      reference
+        "RFC 5492: Capabilities Advertisement with BGP-4, Section
+         4.";
+
+      container mpbgp {
+        when "../../name = 'bt:mp-bgp'";
+        description
+          "Multi-Protocol BGP-specific values.";
+        reference
+          "RFC 4760: Multiprotocol Extensions for BGP-4, Section 8.";
+
+        leaf afi {
+          type iana-rt-types:address-family;
+          description
+           "Address Family Identifier carried in the Multiprotocol
+            Extensions Capability.";
+          reference
+            "RFC 4760: Multiprotocol Extensions for BGP-4, Section
+             8.";
+        }
+        leaf safi {
+          type iana-rt-types:bgp-safi;
+          description
+           "Subsequent Address Family Identifier carried in the
+            Multiprotocol Extensions Capability.";
+          reference
+            "RFC 4760: Multiprotocol Extensions for BGP-4, Section
+             8.";
+        }
+        leaf name {
+          type identityref {
+            base bt:afi-safi-type;
+          }
+          description
+            "When known, name carries the BGP AFI-SAFI identity for
+             the AFI/SAFI combination as used in the BGP YANG
+             modules.";
+        }
+      }
+
+      container graceful-restart {
+        when "../../name = 'bt:graceful-restart'";
+        description
+          "BGP Graceful Restart-specific values.";
+        reference
+          "RFC 4724: Graceful Restart Mechanism for BGP.";
+
+        leaf flags {
+          type bits {
+            /* XXX JMH - how do we show unknown bits? */
```

Comment:
There is a thread started with the yang doctors to try to figure out whether there's precedent for how such things are modeled.

Reply:
A reply from Martin Björklund suggests the pattern of a separate leaf containing bits that cover the unknown entries.  This will require normative text between the two nodes to indicate that unknown bits should only bit present when there's not a registered "known" entry.

We'll also have to figure out if the type we register (since maintenance of the bits likely means moving the type to an IANA-maintained typedef) should include everything we know about even if it's not supported in the main initial module.

Reply:
Adding unknown bit types.  Very noisy, but does the job.  During YANG doctor review, we will want to flag this and consider whether simply registering a common type for unknown flag bits for a given common set of widths make sense to prevent duplication of effort and inconsistent rendering for future users.